statistics/cmsketch.go

```diff
@@ -44,29 +60,188 @@ func NewCMSketch(d, w int32) *CMSketch {
 	return &CMSketch{depth: d, width: w, table: tbl}
 }
 
+// topNHelper wraps some variables used when building cmsketch with top n.
+type topNHelper struct {
+	sampleSize    uint64
+	numTop        uint32
+	counter       map[hack.MutableString]uint64
+	sorted        []uint64
+	onlyOnceItems uint64
+	sumTopN       uint64
+	lastVal       uint64
+}
+
+func newTopNHelper(sample [][]byte, numTop uint32) *topNHelper {
+	counter := make(map[hack.MutableString]uint64)
+	for i := range sample {
+		counter[hack.String(sample[i])]++
+	}
+	sorted, onlyOnceItems := make([]uint64, 0, len(counter)), uint64(0)
+	for _, cnt := range counter {
+		sorted = append(sorted, cnt)
+		if cnt == 1 {
+			onlyOnceItems++
+		}
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] > sorted[j]
+	})
+
+	var (
+		// last is the last element in top N index should occurres atleast `last` times.
+		last      uint64
+		sumTopN   uint64
+		sampleNDV = uint32(len(sorted))
+	)
+	numTop = mathutil.MinUint32(sampleNDV, numTop) // In case numTop is bigger than sample NUV.
+	// The following loop builds find how many elements be added to the top N index.
+	// The final Top-N index may have at most 2*numTop elements for some less skewed sample.
+	for i := uint32(0); i < sampleNDV && i < numTop*2; i++ {
+		// Here, 2/3 is get by running tests, tested 1, 1/2, 2/3, and 2/3 is relative better than 1 and 1/2.
+		// If the frequency of i-th elements is close to n-th element, it is added to the top N index too.
+		if i >= numTop && sorted[i]*3 < sorted[numTop-1]*2 && last != sorted[i] {
+			break
+		}
+		// These two values are only used for build topN, and they are not used in counting defaultValue.
+		last = sorted[i]
+		// We use sumTopN to estimate the total count of top N elements to determine weather to build the top N index.
+		sumTopN += sorted[i]
+	}
+
+	return &topNHelper{uint64(len(sample)), numTop, counter, sorted, onlyOnceItems, sumTopN, last}
+}
+
+// NewCMSketchWithTopN returns a new CM sketch with TopN elements.
+func NewCMSketchWithTopN(d, w int32, sample [][]byte, numTop uint32, rowCount uint64) *CMSketch {
+	helper := newTopNHelper(sample, numTop)
+	estimateNDV, ratio := calculateEstimateNDV(helper, rowCount)
+	c := helper.buildCMSWithTopN(d, w, ratio)
+	c.calculateDefaultVal(helper, estimateNDV, ratio, rowCount)
+	return c
+}
+
+// finalBuild builds Top-N and cmsketch
+func (helper *topNHelper) buildCMSWithTopN(d, w int32, ratio uint64) (c *CMSketch) {
+	c = NewCMSketch(d, w)
+	enableTopN := helper.sampleSize/topNThreshold <= helper.sumTopN
+	topN := make([]dataCount, 0, helper.numTop)
+	helper.sumTopN = 0
+	for counterKey, cnt := range helper.counter {
+		if enableTopN && cnt >= helper.lastVal {
+			topN = append(topN, dataCount{data: hack.Slice(string(counterKey)), count: cnt * ratio})
```

Comment:
extract `cnt*ratio` as a variable, thus we can re-use it in line 131,132 and 134